Adding pynumero build and moving mingw runtime libraries

eslickj opened the pull request:
I'll probably just merge this, but you guys can review it and complain retroactively.  This is hard to test without merging, because to use it from my repo, I need to edit it in a bunch of places to point to a different repo.  I should probably fix that.

eslickj commented:
Looks like I already made it easier to change repos, so I won't merge this until it's done and reviewed.  I'm testing now, but I'm guessing it will need to be tweaked a bit.

eslickj commented:
@jsiirola and @carldlaird I have a slightly stupid question here, but I can't make a decision.  I can't decide if pynumero should go in with solvers or libs.  My current thinking is that solvers are things we'll almost never build once we are happy with them (maybe once a year).  Libs are my property libraries.  They'll change relatively frequently, at least for a while.

I'm leaning toward putting pynumero in solvers.  I know for Pyomo use, you may want solvers and not libs.  I don't want to build solvers frequently though, and I eventually want to do automated builds of libs.

Do you guys have any opinion on where to stick pynumero?  I guess a third tarball is also an option.

jsiirola commented:
I would vote for solvers, only because:
- the pynumero libraries should change very slowly
- `libs` are things that come out of the IDAES codebase, and could (in theory) be built by advanced users.

eslickj commented:
@jsiirola I agree.  As long as the pynumero ASL code rarely changes it's a pretty easy choice.

eslickj commented:
This seems to be working work including pynumero.

eslickj commented:
Seems I can't get pynumero building on centos6 due to C++11 support and the coinbrew script now throws a syntax error only on centos 6.  We may have to give up on centos 6 again.  At least we still have the previous build.

jsiirola commented:
@eslickj: please don't build master.  You want to be building off of Pyomo/pyomo#1381 (from [jsiirola/pyomo:pynumero-testing](https://github.com/jsiirola/pyomo/tree/pynumero-testing)).  That PR has a rework of the libraries / library build harness.  Amonf other things, it eliminates the C++11 requirement.

eslickj commented:
@jsiirola thanks.  I was building the IDAES fork IDAES branch.  I'll switch it over.

carldlaird commented:
@jsiirola Thanks for all this cleanup, John.

eslickj commented:
Okay, with the pynumero-testing version, cento6 build seems to be working.  I'm going to build the rest of the platforms if all goes well I'll have binaries in the dev release soon.  Then I'll make a PR in idaes-dev to switch release URLs, and put all the binaries in bin instead of dividing them between bin and lib.  I'll also need to add a pynumero test or make the IDAES tests run pynumero tests to be sure it's working right.  So far I've only tested the pynumero build on Windows.

michaelbynum commented:
I think we need to be careful distributing AMPL/MP. I believe it has a license (https://github.com/ampl/mp/blob/master/LICENSE.rst). If we want to distribute it, we need to include that license.

eslickj commented:
@michaelbynum, thanks. Will add it.

michaelbynum commented:
@eslickj I think we should get input from @jsiirola first, since he wrote the pynumero build infrastructure. I might be missing something.

eslickj commented:
@michaelbynum ok.  I'm thinking we probably need to add thirdparty license information to the IDAES docs.  There isn't much to see here, but I added the license file and added to the license.txt.  Can't hurt anyway.

eslickj commented on docker/build-windows.ps1:
```diff
@@ -1,7 +1,7 @@
 $buildarg_1 = $args[1]  # just use this to pass in --no-cache or some such
 
-$repo = "https://github.com/idaes/idaes-ext.git"
+$repo = "https://github.com/eslickj/idaes-ext.git"
```
Change this back before merging!

jsiirola commented:
@eslickj: it depends on how you build the pynumero libraries.  The AMPL/MP dependency is not necessary (it is just a clean way to get/build the ASL on all platforms).  Since you are directly using cmake, you can set `-DIPOPT_DIR` to point to the Ipopt install and pick up the ASL from there.  Be sure to *not* set BUILD_AMPLMP_IF_NEEDED to ensure you aren't picking up AMPL/MP.

eslickj commented:
@jsiirola, unless we don't want to include the license info, I don't mind building with AMPL/MP.  The license is pretty similar to a lot of other things we include.  My only hesitation in using ASL from the IPOPT build is that we may want to split the Pynumero build off at some point, so I wanted everything to be independent.  I may be overthinking it. I do build the ASL 3 different times.  I guess it's a bit inefficient.

eslickj commented:
Alright the idaes-dev tests fail with this error.  I'll look into it

```
=================================== FAILURES ===================================

______________________________ test_have_pynumero ______________________________



model = <pyomo.core.base.PyomoModel.ConcreteModel object at 0x7f435366db40>



    def test_have_pynumero(model):

>       nlp = PyomoNLP(model)



idaes/core/tests/test_have_pynumero.py:34: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

/opt/conda/envs/idaes36/lib/python3.6/site-packages/pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:70: in __init__

    super(PyomoNLP, self).__init__(nl_file)

/opt/conda/envs/idaes36/lib/python3.6/site-packages/pyomo/contrib/pynumero/interfaces/ampl_nlp.py:51: in __init__

    self._asl = _asl.AmplInterface(self._nl_file)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 



self = <pyomo.contrib.pynumero.extensions.asl.AmplInterface object at 0x7f435369e4e0>

filename = '/tmp/tmp32lnphxqpynumero.nl', nl_buffer = None



    def __init__(self, filename=None, nl_buffer=None):

    

        if not AmplInterface.available():

            raise RuntimeError(

                "ASL interface is not supported on this platform (%s)"

>               % (os.name,) )

E           RuntimeError: ASL interface is not supported on this platform (posix)



/opt/conda/envs/idaes36/lib/python3.6/site-packages/pyomo/contrib/pynumero/extensions/asl.py:33: RuntimeError

=========================== short test summary info ============================

FAILED idaes/core/tests/test_have_pynumero.py::test_have_pynumero - RuntimeEr...

==== 1 failed, 2514 passed, 12 skipped, 18 deselected in 362.30s (0:06:02) =====


Exited with code exit status 1

CircleCI received exit code 1
```

jsiirola commented:
Where are you installing `libpynumero_ASL.so` to?  Is that location either a standard system location or else listed on `LD_LIBRARY_PATH`?  What platform are you running on (what is the result of `platform.system()`)?

eslickj commented:
@jsiirola it's in the LD_LIBRARY_PATH.  I'm still working on tracking it down, but I tried the utility function (find_pynumero_library) and it does seem to find the library.  Still working working widening the net.

jsiirola commented:
@eslickj: what do you get when running `platform.system()`?

eslickj commented:
@jsiirola,

```
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.system()
'Linux'
```

eslickj commented:
@jsiirola, I have a theory the AmplInterface class starts like this.

```
class AmplInterface(object):

    libname = find_pynumero_library('pynumero_ASL') 
```

I guess that ```libname``` is being set before importing idaes updates LD_LIBRARY_PATH.  So when I call the find library function later it works, but pynumero is stuck with libname.

jsiirola commented:
Ick. I think you are correct.  I can pt a patch in to the PR to defer resolving the library...

jsiirola commented:
@eslickj: Patch is in; try again?

eslickj commented:
Yes.  Working on it now.

eslickj commented:
@jsiirola, okay looks like that solved it.  It's working now.

eslickj commented:
Ran into this problem on the Ubuntu 19.10build (and I assume the 20.04 build), so maybe I'll use Ipopt's ASL build after all.

```
/repo/idaes-ext/pyomo/pyomo/contrib/pynumero/src/build/amplmp-prefix/src/amplmp/thirdparty/benchmark/src/benchmark.cc: In copy constructor 'benchmark::internal::Benchmark::Benchmark(const benchmark::internal::Benchmark&)':
/repo/idaes-ext/pyomo/pyomo/contrib/pynumero/src/build/amplmp-prefix/src/amplmp/thirdparty/benchmark/src/benchmark.cc:525:38: error: implicitly-declared 'benchmark::internal::BenchmarkImp::BenchmarkImp(const benchmark::internal::BenchmarkImp&)' is deprecated [-Werror=deprecated-copy]
  525 |   : imp_(new BenchmarkImp(*other.imp_))
      |                                      ^
/repo/idaes-ext/pyomo/pyomo/contrib/pynumero/src/build/amplmp-prefix/src/amplmp/thirdparty/benchmark/src/benchmark.cc:320:17: note: because 'benchmark::internal::BenchmarkImp' has user-provided 'benchmark::internal::BenchmarkImp& benchmark::internal::BenchmarkImp::operator=(const benchmark::internal::BenchmarkImp&)'
  320 |   BenchmarkImp& operator=(BenchmarkImp const&);
      |                 ^~~~~~~~
cc1plus: all warnings being treated as errors
```

eslickj commented:
Alright I changed the pynumero build to use Ipopt's ASL build and it seems to work fine.  Now I'll rebuild and retest.  If everything works, I'll go ahead and merge this and put the binaries on the dev release.

I also changed the solver build script to allow you to build everything without the HSL if you don't have the files.  That may come in handy for public builds and tests.  I'm not quite sure where I going with this.

eslickj commented:
@jsiirola, I had a little trouble with cmake on windows with msys2 and mingw64.  For some weird reason I had to specify the ASL_INCLUDE_DIR variable like so

```
cmake -DIPOPT_DIR=$IDAES_EXT/coinbrew/dist -DASL_INCLUDE_DIR=$IDAES_EXT/coinbrew/dist/include -G"MSYS Makefiles" ..
``` 

I also had to just remove the INSTALL command from CMakeLists.txt, because I was getting this error

```
INSTALL Library TARGETS given no DESTINATION!
```

It's easy enough to work around, and it could be my fault that something isn't quite set up right, but I figured it may be of some interest to you.  Everything was totally fine on Linux, but like I said this really doesn't cause me any trouble so it's not a big deal.

jsiirola commented:
@eslickj: what directory layout under `dist/include` does coinbrew create?  Older build harnesses had a number of subdirectories (`coin`, `coin/ThirdParty`, `cppad`).  Is that all flattened now?

eslickj commented:
@jsiirola, I didn't check on Linux because it just worked.  On Windows it was ```dist/include/coin-or/asl``` where I found the ASL header files.  Not sure why it would be different on Windows, if it is.  Once my builds finish on docker, I can go back in and look closer if you need me to.

jsiirola commented:
Yes, please do.  I can add additional directories to look in under `IPOPT_DIR` so that you don't have to specify both `IPOPT_DIR` and `ASL_INCLUDE_DIR`.  Also, can you check if modifying the INSTALL command to:
```cmake
INSTALL(TARGETS pynumero_ASL
    LIBRARY DESTINATION lib
    RUNTIME DESTINATION lib
)
```
works everywhere?  If it does, I will update the Pyomo PR.

eslickj commented:
@jsiirola, I'm about the check cmake on centos6, but the include directory structure is the same on CentOS6 as it was on Windows.  Not sure if the search functions are somehow different on windows and maybe it didn't look all the way down the tree or what.  The .a library file was found okay on both.

Okay, I'm, checking the CMakeList.txt modification now.  Let me know if there is anything else that would be helpful.

jsiirola commented:
I think it is that on Linux the ASL headers were being picked up elsewhere (or, more likely, through pkg-config).  The `include/coin-or/asl` subdirectory is new, and I will update the CMakeLists to also look there.

jsiirola commented:
I updated the Pyomo PR with both of the requested changes and would be interested to know if that resolves the problems everywhere.

eslickj commented:
@jsiirola, I figure it should.  I'll take the work-arounds out of the build scripts.  And rebuild.  I'm building off your PR, so I'll let you know if I encounter any trouble.  It will take a while to get through all 7 builds, but I don't anticipate anything going wrong.

eslickj commented:
@jsiirola, so far windows and coentos6 work as expected.  The rest probably will be fine to but I'll get them rebuilt too to be sure.

eslickj commented:
Okay seems to be working.  I'm going to go ahead and merge.
